Replumbing: add a configuration module for providers

levitte opened the pull request:
This configuration module supports a configuration structure pretty
much like the engine configuration module, i.e. something like this:

    openssl_conf = openssl_init

    [openssl_init]
    providers = provider_section

    [provider_section]
    # Configure the provider named "foo"
    foo = foo_section
    # Configure the provider named "bar"
    bar = bar_section

    [foo_section]
    # Override name given in the provider section
    identity = myfoo
    # The exact path of the module.  This is platform specific
    module_path = /opt/openssl/modules/foo.so
    # Whether it should be automatically activated.  Value is unimportant
    activate = whatever

    # Anything else goes as well, and becomes parameters that the
    # provider can get
    what = 1
    # sub-sections will be followed as well
    ever = ever_section

    [ever_section]
    cookie = monster

All the configurations in a provider section and its sub-sections
become parameters for the provider to get, i.e. the "foo" provider
will be able to get values for the following keys (with associated
values shown):

    identity            => myfoo
    module_path         => /opt/openssl/modules/foo.so
    activate            => whatever
    what                => 1
    ever.cookie         => monster

<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING

Other than that, provide a description above this comment if there isn't one already

If this fixes a github issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ ] documentation is added or updated
- [ ] tests are added or updated

levitte committed:
Replumbing: add a configuration module for providers

This configuration module supports a configuration structure pretty
much like the engine configuration module, i.e. something like this:

    openssl_conf = openssl_init

    [openssl_init]
    providers = provider_section

    [provider_section]
    # Configure the provider named "foo"
    foo = foo_section
    # Configure the provider named "bar"
    bar = bar_section

    [foo_section]
    # Override name given in the provider section
    identity = myfoo
    # The exact path of the module.  This is platform specific
    module_path = /opt/openssl/modules/foo.so
    # Whether it should be automatically activated.  Value is unimportant
    activate = whatever

    # Anything else goes as well, and becomes parameters that the
    # provider can get
    what = 1
    # sub-sections will be followed as well
    ever = ever_section

    [ever_section]
    cookie = monster

All the configurations in a provider section and its sub-sections
become parameters for the provider to get, i.e. the "foo" provider
will be able to get values for the following keys (with associated
values shown):

    identity            => myfoo
    module_path         => /opt/openssl/modules/foo.so
    activate            => whatever
    what                => 1
    ever.cookie         => monster

levitte committed:
Replumbing: add functionality to set provider parameters

Provider parameters are parameters set by the core that the provider
can retrieve.  The primary use it to support making OpenSSL
configuration data available to the provider.

slontis commented:
I was thinking that the self test variables would be in a separate config file that gets saved during the install process.. Ideally this would be loaded by the core also (A config file loaded from the config file).

levitte commented:
That could be done using the new `.include` directive.

That separate config file will need to contain a section name.

levitte commented:
So @slontis, what I'm suggesting is that the main config file could contain something like this:

```
openssl_conf = openssl_init

[openssl_init]
providers = provider_section

[provider_section]
fips = fips_module_section

[fips_module_section]
identity = fips
module_path = /opt/openssl/modules/fips.so
activate = whatever
fips_chk = fips_check_section

.include /opt/openssl/modules/fips.conf
```

And then, `/opt/openssl/modules/fips.conf` could contain this:

```
[fips_check_section]
module.mac = DCBD40AEC65D0E62B93E316F9315741021911D2ACEAFEF40730546845E83B90D
install.mac = FE10135AD3B4C7821B1E174CAC840C0AEF5F0696AE6E2C850E8E552886D62AD9
install.status = INSTALL_SELF_TEST_KATS_RUN
```

The resulting param names that the fips module would have to retrieve are `fips_chk.module.mac`, `fips_chk.install.mac` and `fips_chk.install.status`.

-----

Another variant is just this in the main config file:

```
openssl_conf = openssl_init

[openssl_init]
providers = provider_section

[provider_section]
fips = fips_module_section

.include /opt/openssl/modules/fips.conf
```

And the rest in `/opt/openssl/modules/fips.conf` could contain this:

```
[fips_module_section]
identity = fips
module_path = /opt/openssl/modules/fips.so
activate = whatever
module.mac = DCBD40AEC65D0E62B93E316F9315741021911D2ACEAFEF40730546845E83B90D
install.mac = FE10135AD3B4C7821B1E174CAC840C0AEF5F0696AE6E2C850E8E552886D62AD9
install.status = INSTALL_SELF_TEST_KATS_RUN
```

That would make for exactly the param names you talk about instead of having to deal with the `fips_chk` prefix.

-----

Either way we choose, I believe this PR supports it.

levitte commented:
Close / re-open to kick the CIs

levitte commented:
This is now open for review

slontis commented:
The install application needs to write out these config values (which is why we decided that a seperate file should be used).

levitte commented:
> The install application needs to write out these config values (which is why we decided that a seperate file should be used).

Yes?  Can I assume that this is basically a mental note?

slontis commented:
yes :)

slontis commented on crypto/provider_conf.c:
```diff
@@ -0,0 +1,168 @@
+
+static void provider_conf_finish(CONF_IMODULE *md)
+{
+#if 0
```
#if 0 ?

levitte replied:
Thanks, will fix

levitte commented:
I took this out of draft mode a bit too early...  so making it WIP

levitte committed:
Add test for the provider configuration module

We reuse test/provider_internal_test.c and test/p_test.c,
and get it loaded one more time via the configuration file
test/provider_internal_test.conf

To support different platform standards regarding module
extensions, we generate test/provider_internal_test.conf

levitte commented:
Now also includes a test

levitte committed:
Replumbing: add documentation for the provider configuration module

slontis commented on crypto/provider_conf.c:
```diff
@@ -0,0 +1,170 @@
+
+    if (activate) {
+        ossl_provider_activate(prov);
+        if (activated_providers == NULL)
```
Is activated_providers thread safe?

levitte replied:
This is all governed by `OPENSSL_init_crypto`, is thread safety in config modules relevant?  If it is, we'd better review *all* our config modules.

slontis commented on crypto/provider_conf.c:
```diff
@@ -0,0 +1,170 @@
+}
+
+
+static void provider_conf_finish(CONF_IMODULE *md)
```
provider_conf_deinit

levitte replied:
Sure

slontis commented on doc/man5/config.pod:
```diff
@@ -216,6 +221,57 @@ For example:
+
+The command B<identity> is used to give the ENGINE name. For example:
+
+ [engine_section]
```
This is to handle engines which will be deprecated at some time in the future?

levitte replied:
Oops, this is a copy paste error.

Yes, the intention is that engines will eventually go away

levitte commented:
The tests here are currently failing, and it's due to confusion on when modules exist or not.  #8623 helps a lot with that.

levitte commented:
Finally hammered this through the CIs...

levitte commented:
Final review?

paulidale commented on crypto/provider_conf.c:
```diff
@@ -0,0 +1,178 @@
+        for (i = 0; i < sk_CONF_VALUE_num(sect); i++) {
+            CONF_VALUE *sectconf = sk_CONF_VALUE_value(sect, i);
+
+            if (buffer_len + strlen(sectconf->name) >= sizeof(buffer))
```
The '\0' terminator would add one. wouldn't it?

levitte replied:
You do notice `>=`, right?

paulidale replied:
I did but didn't think it through far enough.

paulidale commented on crypto/provider_conf.c:
```diff
@@ -0,0 +1,178 @@
+            if (buffer_len + strlen(sectconf->name) >= sizeof(buffer))
+                return 0;
+            buffer[buffer_len] = '\0';
+            OPENSSL_strlcat(buffer, sectconf->name, sizeof(buffer));
```
memcpy(3) or strcpy(3) would work as well here but this is arguably safer.

Got to give static analysers some false positives :)

paulidale commented on crypto/provider_conf.c:
```diff
@@ -0,0 +1,178 @@
+{
+    STACK_OF(CONF_VALUE) *elist;
+    CONF_VALUE *cval;
+    int i;
```
Blank line after this.

levitte replied:
Fixed

paulidale commented on crypto/provider_core.c:
```diff
@@ -278,6 +296,28 @@ void ossl_provider_free(OSSL_PROVIDER *prov)
+/* Setters */
+int ossl_provider_set_module_path(OSSL_PROVIDER *prov, const char *module_path)
+{
+    prov->path = OPENSSL_strdup(module_path);
```
Check for allocation failure?  Otherwise it might as well be a void function.

levitte replied:
Also, if there was a previous value, it will be leaked

levitte replied:
Fixed

paulidale commented on crypto/provider_core.c:
```diff
@@ -278,6 +296,28 @@ void ossl_provider_free(OSSL_PROVIDER *prov)
+int ossl_provider_add_parameter(OSSL_PROVIDER *prov,
+                                const char *name, const char *value)
+{
+    INFOPAIR *pair = OPENSSL_malloc(sizeof(*pair));
```
Missing NULL check after this allocation.

levitte replied:
Fixed

paulidale commented on crypto/provider_core.c:
```diff
@@ -278,6 +296,28 @@ void ossl_provider_free(OSSL_PROVIDER *prov)
+    INFOPAIR *pair = OPENSSL_malloc(sizeof(*pair));
+
+    if (prov->parameters == NULL)
+        prov->parameters = sk_INFOPAIR_new_null();
```
NULL check here?

levitte replied:
Fixed

paulidale commented on crypto/provider_core.c:
```diff
@@ -278,6 +296,28 @@ void ossl_provider_free(OSSL_PROVIDER *prov)
+        prov->parameters = sk_INFOPAIR_new_null();
+
+    pair->name = OPENSSL_strdup(name);
+    pair->value = OPENSSL_strdup(value);
```
NULL checks for these two too.

levitte replied:
Fixed

paulidale commented on crypto/provider_core.c:
```diff
@@ -278,6 +296,28 @@ void ossl_provider_free(OSSL_PROVIDER *prov)
+
+    pair->name = OPENSSL_strdup(name);
+    pair->value = OPENSSL_strdup(value);
+    sk_INFOPAIR_push(prov->parameters, pair);
```
Failure check here?

levitte replied:
Fixed

paulidale commented on doc/man5/config.pod:
```diff
@@ -216,6 +221,57 @@ For example:
+value of this variable points to a section containing further provider
+configuration information.
+
+The section pointed to by B<providers> is a table of engine names
```
Engine names?  If we're trying to move away from engines, it might be better to write _provider names_.

levitte replied:
Specially since this is about providers! :wink:

levitte replied:
Fixed

paulidale commented on doc/man5/config.pod:
```diff
@@ -216,6 +221,57 @@ For example:
+ # Override default name and use "myfoo" instead.
+ identity = myfoo
+
+The parameters B<module> loads and adds a provider module from the
```
I think this would read better as a singular _parameter_.

levitte replied:
Fixed

paulidale commented on doc/man5/config.pod:
```diff
@@ -216,6 +221,57 @@ For example:
+given module path.  That path may be a simple file name, a relative
+path or an absolute path.
+
+The parameters B<activate> determines whether to activate the
```
Again, a singular _parameter_.

levitte replied:
Fixed

paulidale commented on test/p_test.c:
```diff
@@ -52,21 +52,32 @@ static int p_get_params(const OSSL_PROVIDER *prov, OSSL_PARAM params[])
+                } else {
+                    const char *versionp = *(void **)counter_request[0].data;
+                    const char *namep = *(void **)counter_request[1].data;
+                    sprintf(buf, "Hello OpenSSL %.20s, greetings from %s!",
```
Blank line before this.

levitte replied:
Fixed

paulidale reviewed (changes requested):
Looks good, mostly nits here but a possible off by one and some missing error return checks.

levitte commented:
All things answered, I think...  including CI failures

paulidale commented on crypto/provider_core.c:
```diff
@@ -278,6 +296,40 @@ void ossl_provider_free(OSSL_PROVIDER *prov)
+        OPENSSL_free(pair->name);
+        OPENSSL_free(pair->name);
+    }
+    OPENSSL_free(pair);
```
Inside the if statement?

The NULL check has been paid for already.

levitte replied:
Yeahok

paulidale reviewed (approved):
Looks good.

One comment which is kind of stylistic so keep or address as you feel like.

slontis commented on crypto/provider_core.c:
```diff
@@ -278,6 +296,40 @@ void ossl_provider_free(OSSL_PROVIDER *prov)
+
+    if (pair != NULL) {
+        OPENSSL_free(pair->name);
+        OPENSSL_free(pair->name);
```
I think you want pair->value here

levitte replied:
Ohhhhhh, good catch!

slontis replied:
or add the null check to free_infopair() and just call that.

levitte replied:
Fixed

levitte committed:
fixup! Replumbing: add functionality to set provider parameters

levitte commented:
Merged.  Thankss for all the reviews!

b6670f690c4244e63dbc02a2ba25061f9c53945f Replumbing: add documentation for the provider configuration module
6d872a838df78518508b5661d98da62a097317b1 Add test for the provider configuration module
abbc2c408385326d9c9cd60b92a6c92b945c1d96 Replumbing: add a configuration module for providers
ac1055ef13ccb5789e2bed7b9688c8eb16dd05ce Replumbing: add functionality to set provider parameters
